Mage.Sets/src/mage/cards/t/TheBalrogOfMoria.java

```diff
@@ -0,0 +1,100 @@
+package mage.cards.t;
+
+import mage.MageInt;
+import mage.abilities.Ability;
+import mage.abilities.common.CycleTriggeredAbility;
+import mage.abilities.common.DiesSourceTriggeredAbility;
+import mage.abilities.common.delayed.ReflexiveTriggeredAbility;
+import mage.abilities.costs.common.ExileSourceFromGraveCost;
+import mage.abilities.costs.mana.ManaCostsImpl;
+import mage.abilities.effects.common.CreateTokenEffect;
+import mage.abilities.effects.common.DoWhenCostPaid;
+import mage.abilities.effects.common.ExileTargetEffect;
+import mage.abilities.keyword.CyclingAbility;
+import mage.abilities.keyword.HasteAbility;
+import mage.abilities.keyword.TrampleAbility;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.CardType;
+import mage.constants.SubType;
+import mage.constants.SuperType;
+import mage.filter.FilterPermanent;
+import mage.filter.common.FilterCreaturePermanent;
+import mage.filter.predicate.permanent.ControllerIdPredicate;
+import mage.game.Game;
+import mage.game.permanent.token.TreasureToken;
+import mage.players.Player;
+import mage.target.TargetPermanent;
+import mage.target.targetadjustment.TargetAdjuster;
+import mage.target.targetpointer.EachTargetPointer;
+
+import java.util.UUID;
+
+/**
+ *
+ * @author Susucr
+ */
+public final class TheBalrogOfMoria extends CardImpl {
+
+    public TheBalrogOfMoria(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{4}{B}{B}{R}");
+        
+        this.supertype.add(SuperType.LEGENDARY);
+        this.subtype.add(SubType.AVATAR);
+        this.subtype.add(SubType.DEMON);
+        this.power = new MageInt(8);
+        this.toughness = new MageInt(8);
+
+        // Trample
+        this.addAbility(TrampleAbility.getInstance());
+
+        // Haste
+        this.addAbility(HasteAbility.getInstance());
+
+        // When The Balrog of Moria dies, you may exile it. When you do, for each opponent, exile up to one target creature that player controls.
+        ReflexiveTriggeredAbility reflexiveAbility = new ReflexiveTriggeredAbility(
+            new ExileTargetEffect()
+                .setTargetPointer(new EachTargetPointer())
```

Comment:
Something is weird with `ExileTargetEffect()` and that manually set `TargetPointer`.

When testing the card with multiple opponent, each with a valid creature to target,
the ability is properly asking for a target per opponent.

However in ExileTargetExile.java, we have the following loop over the targets:

```@Override
    public boolean apply(Game game, Ability source) {
        Player controller = game.getPlayer(source.getControllerId());
        if (controller != null) {
            Set<Card> toExile = new LinkedHashSet<>();
            for (UUID targetId : getTargetPointer().getTargets(game, source)) {```

That `getTargetPointer()` is not the `EachTargetPointer()` I was expecting (it is a `FirstTargetPointer` with a single target). `source.getTargets()` has all the targetted creatures under it.

I am not sure what is wrong there, is it a bug where a wrong `TargetPointer` is copied at some point, or is the way I did set up the targetting wrong in the balrog's code?

Reply:
My guess is that `.setTargetPointer(new EachTargetPointer())` has to be on the `DoWhenCostPaid` which copies its target pointer to the reflexive ability

Reply:
thanks, I'll test that instead.

Reply:
That was it. Now the multiple targets are properly exiled.